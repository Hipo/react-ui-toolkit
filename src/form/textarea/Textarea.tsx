import "./_textarea.scss";

import React, {useState} from "react";
import classNames from "classnames";
import TextareaAutosize, {TextareaAutosizeProps} from "react-textarea-autosize";

import {KEYBOARD_EVENT_KEY} from "../../core/utils/keyboard/keyboardEventConstants";

export type TextareaProps = Omit<
  React.TextareaHTMLAttributes<HTMLTextAreaElement>,
  "name" | "onChange" | "className"
> & {
  name: string;
  onChange: React.ReactEventHandler<HTMLTextAreaElement>;
  value?: string;
  testid?: string;
  isDisabled?: boolean;
  customClassNames?: {
    container?: string;
    textarea?: string;
  };
  style?: React.CSSProperties;
  autoSizeProps?: TextareaAutosizeProps;
  onJustEnterPressed?: () => void;
  onShiftEnter?: () => void;
  isRequired?: boolean;
  hasError?: boolean;
};

function Textarea(props: TextareaProps) {
  const {
    name,
    value,
    isDisabled,
    customClassNames,
    testid,
    onKeyUp,
    style,
    autoSizeProps,
    onJustEnterPressed,
    isRequired,
    onShiftEnter,
    hasError,
    onKeyDown,
    autoComplete = "off",
    autoCorrect = "off",
    autoFocus = false,
    ...otherProps
  } = props;
  const [isShiftPressed, setShiftPressedState] = useState(false);
  const containerClassName = classNames(
    "textarea-container",
    customClassNames?.container,
    {
      "textarea--has-error": hasError,
      "textarea--is-disabled": isDisabled
    }
  );
  const textareaClassName = classNames("textarea", customClassNames?.textarea);

  return (
    <div className={containerClassName}>
      {autoSizeProps ? (
        <TextareaAutosize
          className={textareaClassName}
          data-testid={testid}
          name={name}
          value={value}
          autoComplete={autoComplete}
          autoCorrect={autoCorrect}
          disabled={isDisabled}
          onKeyDown={handleKeyDown}
          onKeyUp={handleKeyUp}
          style={autoSizeProps.style}
          maxRows={autoSizeProps.maxRows}
          minRows={autoSizeProps.minRows}
          onHeightChange={autoSizeProps.onHeightChange}
          required={isRequired}
          autoFocus={autoFocus}
          {...otherProps}
        />
      ) : (
        <textarea
          className={textareaClassName}
          data-testid={testid}
          name={name}
          value={value}
          autoComplete={autoComplete}
          autoCorrect={autoCorrect}
          disabled={isDisabled}
          onKeyDown={handleKeyDown}
          onKeyUp={handleKeyUp}
          style={style}
          required={isRequired}
          autoFocus={autoFocus}
          {...otherProps}
        />
      )}
    </div>
  );

  function handleKeyDown(event: React.KeyboardEvent<HTMLTextAreaElement>) {
    const {key: pressedKey} = event;
    const newEvent = {...event};

    newEvent.currentTarget.name = name;

    if (pressedKey === KEYBOARD_EVENT_KEY.ENTER) {
      if (!isShiftPressed && onJustEnterPressed) {
        event.preventDefault();
        onJustEnterPressed();
      } else if (isShiftPressed && onShiftEnter) {
        event.preventDefault();
        onShiftEnter();
      }
    } else if (pressedKey === KEYBOARD_EVENT_KEY.SHIFT) {
      setShiftPressedState(true);
    }

    if (onKeyDown) {
      onKeyDown(newEvent);
    }
  }

  function handleKeyUp(event: React.KeyboardEvent<HTMLTextAreaElement>) {
    const {key: pressedKey} = event;

    if (pressedKey === KEYBOARD_EVENT_KEY.SHIFT) {
      setShiftPressedState(false);
    }

    if (onKeyUp) {
      onKeyUp(event);
    }
  }
}

export default Textarea;
