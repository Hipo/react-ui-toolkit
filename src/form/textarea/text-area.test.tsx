import React from "react";
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../core/utils/test/testUtils";
import Textarea, {TextareaProps} from "./Textarea";

describe("<TextArea />", () => {
  afterEach(cleanup);

  const defaultTextAreaProps: TextareaProps = {
    testid: "text-area",
    name: "text-area",
    onChange: jest.fn()
  };

  it("should render correctly", () => {
    render(<Textarea {...defaultTextAreaProps} />);
  });

  it("should matches snapshot", () => {
    const tree = create(<Textarea {...defaultTextAreaProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<Textarea {...defaultTextAreaProps} />);

    await testA11y(container, {rules: {label: {enabled: false}}});
  });

  it("keyboard events should work correctly", () => {
    const handleShiftEnter = jest.fn();
    const handleJustEnterPressed = jest.fn();
    const handleKeyDown = jest.fn();
    const handleKeyUp = jest.fn();

    const {getByTestId} = render(
      <Textarea
        onKeyUp={handleKeyUp}
        onKeyDown={handleKeyDown}
        onJustEnterPressed={handleJustEnterPressed}
        onShiftEnter={handleShiftEnter}
        {...defaultTextAreaProps}
      />
    );
    const textArea = getByTestId("text-area");

    fireEvent.keyDown(textArea, {
      key: "Shift",
      keyCode: 16,
      code: "ShiftLeft"
    });
    fireEvent.keyUp(textArea, {
      key: "Shift",
      keyCode: 16,
      code: "ShiftLeft"
    });
    fireEvent.keyDown(textArea, {
      key: "Enter",
      code: "Enter",
      keyCode: 13
    });

    expect(handleShiftEnter).not.toHaveBeenCalled();
    expect(handleJustEnterPressed).toHaveBeenCalledTimes(1);
    // eslint-disable-next-line no-magic-numbers
    expect(handleKeyDown).toHaveBeenCalledTimes(2);
    expect(handleKeyUp).toHaveBeenCalledTimes(1);

    jest.clearAllMocks();

    fireEvent.keyDown(textArea, {
      key: "Shift",
      keyCode: 16,
      code: "ShiftLeft"
    });
    fireEvent.keyDown(textArea, {
      key: "Enter",
      code: "Enter",
      keyCode: 13
    });

    expect(handleShiftEnter).toHaveBeenCalledTimes(1);
    expect(handleJustEnterPressed).not.toHaveBeenCalled();
    // eslint-disable-next-line no-magic-numbers
    expect(handleKeyDown).toHaveBeenCalledTimes(2);
  });

  it("should have proper class name", () => {
    const {getByTestId} = render(
      <Textarea
        customClassNames={{
          container: "text-area__container",
          textarea: "text-area__textarea"
        }}
        {...defaultTextAreaProps}
      />
    );

    expect(getByTestId("textarea-container")).toHaveClass("text-area__container");
    expect(getByTestId("text-area")).toHaveClass("text-area__textarea");
  });
});
