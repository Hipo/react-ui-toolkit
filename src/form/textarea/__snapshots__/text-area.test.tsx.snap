// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should matches snapshot 1`] = `
<div
  className="textarea-container"
  data-testid="textarea-container"
>
  <textarea
    autoComplete="off"
    autoCorrect="off"
    autoFocus={false}
    className="textarea"
    data-testid="text-area"
    name="text-area"
    onChange={[MockFunction]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
</div>
`;
